### PR TITLE
Fix NullPointerException when jib-maven-plugin is configured with environment variables

### LINKED ISSUE #632: NullPointerException  when jib-maven-plugin is configured with environment variables
**Description of the issue**:

`NullPointerException` when `jib-maven-plugin` is configured with environment variables.

**Expected behavior**:

Normal image to be built.

**Steps to reproduce**:

`mvn jib:dockerBuild`

**Environment**:

Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T19:33:14+01:00)
Maven home: C:\Apps\apache-maven-3.5.4
Java version: 1.8.0_172, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk1.8.0_172\jre
Default locale: en_GB, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

**`jib-maven-plugin` Configuration**:

```xml
<plugin>
  <groupId>com.google.cloud.tools</groupId>
  <artifactId>jib-maven-plugin</artifactId>
  <version>0.9.7-SNAPSHOT</version>
  <configuration>
    <environment>
      <SPRING_CONFIG_LOCATION>file:/path/application.properties</SPRING_CONFIG_LOCATION>
    </environment>
  </configuration>
  ...
</plugin>
```

**Log output**:

```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.7-SNAPSHOT:dockerBuild (default-cli) on project mig-management-portal: Execution default-cli of goal com.google.cloud.tools:jib-maven-plugin:0.9.7-SNAPSHOT:dockerBuild failed. NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.7-SNAPSHOT:dockerBuild (default-cli) on project mig-management-portal: Execution default-cli of goal com.google.cloud.tools:jib-maven-plugin:0.9.7-SNAPSHOT:dockerBuild failed.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:60)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal com.google.cloud.tools:jib-maven-plugin:0.9.7-SNAPSHOT:dockerBuild failed.
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:60)
Caused by: java.lang.NullPointerException
    at java.util.TreeMap.getEntry (TreeMap.java:347)
    at java.util.TreeMap.containsKey (TreeMap.java:232)
    at com.google.cloud.tools.jib.builder.BuildConfiguration$Builder.setEnvironment (BuildConfiguration.java:121)
    at com.google.cloud.tools.jib.maven.BuildDockerMojo.execute (BuildDockerMojo.java:92)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:60)
```

**Additional Information**: 

Maven provides the environment `Map` as a `TreeMap`. When jib attempts to verify that there are no `null` keys, `TreeMap` throws the `NullPointerException`. `TreeMap.containsKey(key)` can't accept `null` as a parameter.

The unit test uses an `ImmutableMap.of()` to provide the environment, which provides a `SingletonImmutableBiMap`, whose `containsKey()` method has the `@Nullable` annotation on the `key` parameter.

`TreeMap`s javadoc for `containsKey()` says: "@throws NullPointerException if the specified key is null and this map uses natural ordering, or its comparator does not permit null keys"

Replacing the `ImmableMap.of()` in the test with a `TreeMap` instance produces the same error as above.

I'm happy to look at this and propose a solution as a pull request.

### PR DESCRIPTION
**Pull request by d5nguyenvan**
Fix #632:
* Use `Iterables.any(environmentMap.keySet(), Objects::isNull)` instead of `environmentMap.containsKey(null)`
* Use `!Iterables.any(environmentMap.values(), Objects::isNull)` instead of `environmentMap.containsValue(null)`
* Add case to test null value for `setJavaArguments`, `setJvmFlags`, `setExposedPorts`, `setEnvironment`

### CODE REVIEW

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java [hunk 1]
```diff
@@ -200,6 +202,69 @@ public void testBuilder_missingValues() {
...
+    try {
+      BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(environmentMap);
+      Assert.fail("The IllegalArgumentException should be thrown.");
+    } catch (IllegalArgumentException ex) {
+      Assert.assertNull(ex.getMessage());
+    }
+
+    // Environment map can accept TreeMap and Hashtable
+    try {
+      BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(Maps.newTreeMap());
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java]: Let's do `new TreeMap<>()`.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java [hunk 2]
```diff
@@ -200,6 +202,69 @@ public void testBuilder_missingValues() {
...
+
+    try {
+      BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(environmentMap);
+      Assert.fail("The IllegalArgumentException should be thrown.");
+    } catch (IllegalArgumentException ex) {
+      Assert.assertNull(ex.getMessage());
+    }
+
+    // Environment values element should not be null.
+    environmentMap = Maps.newHashMap();
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java]: Let's not reuse an existing variable. `Map<String, String> nullValueMap = ...`?

**Comment by coollog** [comment]
@d5nguyenvan Thanks for contributing! This looks great :) I'm going to defer merge approval to another member of the @GoogleContainerTools/java-tools team

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java [hunk 3]
```diff
@@ -200,6 +202,69 @@ public void testBuilder_missingValues() {
...
+      Assert.fail("The IllegalArgumentException should be thrown.");
+    } catch (IllegalArgumentException ex) {
+      Assert.assertNull(ex.getMessage());
+    }
+
+    // Can accept empty environment
+    BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(ImmutableMap.of());
+
+    // Environment keys element should not be null.
+    Map<String, String> environmentMap = Maps.newHashMap();
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java]: Let's do `new HashMap<>()`.

> Note for Java 7 and later: this method is now unnecessary and should be treated as deprecated. Instead, use the HashMap constructor directly, taking advantage of the new "diamond" syntax.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java [hunk 4]
```diff
@@ -200,6 +202,69 @@ public void testBuilder_missingValues() {
...
+      Assert.fail("The IllegalArgumentException should be thrown.");
+    } catch (IllegalArgumentException ex) {
+      Assert.assertNull(ex.getMessage());
+    }
+
+    // Environment map can accept TreeMap and Hashtable
+    try {
+      BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(Maps.newTreeMap());
+      BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(new Hashtable<>());
+    } catch (NullPointerException ex) {
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java]: I see above you did not catch the exception.
```
    // Can accept empty environment
    BuildConfiguration.builder(Mockito.mock(BuildLogger.class)).setEnvironment(ImmutableMap.of());
```
Let's be consistent and remove this try-catch block, throwing potential NPEs to the test framework. (Moreover, if you were to catch, you should wrap individual calls to `setEnvironment()`.)

BTW, you'd normally do `Assert.fail()` instead of directly doing `throw new AssertionError()`.

Comment (d5nguyenvan) [reply jib-core/src/test/java/com/google/cloud/tools/jib/builder/BuildConfigurationTest.java]: Yup, let me revise the test, however, for `Assert.fail()` I did that and it violate the error-prone checking so that why I change to `throw new AssertionError()`
